cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/server.go

```diff
@@ -168,20 +178,53 @@ func getHelmRepos(ctx context.Context) (*unstructured.UnstructuredList, error) {
 		// TODO: should we filter out those repos that don't have .status.condition.Ready == True?
 		// like we do in GetAvailablePackages()?
 		// i.e. should GetAvailableRepos() call semantics be such that only "Ready" repos are returned
+		// ongoing slack discussion https://vmware.slack.com/archives/C4HEXCX3N/p1621846518123800
 		return repos, nil
 	}
 }
 
-func readPackagesFromRepoIndex(indexURL string) ([]*corev1.AvailablePackage, error) {
-	//Get the response bytes from the url
+func readPackagesFromRepoIndex(repoRef *corev1.AvailablePackage_PackageRepositoryReference, indexURL string) ([]*corev1.AvailablePackage, error) {
+	index, err := getHelmIndexFileFromURL(indexURL)
+	if err != nil {
+		return nil, err
+	}
+
+	responsePackages := []*corev1.AvailablePackage{}
+	for _, entry := range index.Entries {
+		// note that entry itself is an array of chart versions
+		// after index.SortEntires() call, it looks like there is only one entry per package,
+		// and entry[0] should be the most recent chart version, e.g. Name: "mariadb" Version: "9.3.12"
+		// while the rest of the elements in the entry array keep track of all chart versions, e.g.
+		// "mariadb" version "9.3.11", "9.3.10", etc. For entry "mariadb", bitnami catalog has
+		// almost 200 chart versions going all the way back many years to version "2.1.4".
+		// So for now, let's just keep track of the latest, not to overwhelm the caller with
+		// all these outdated versions
+		if entry[0].GetDeprecated() {
+			log.Infof("skipping deprecated chart: [%s]", entry[0].Name)
+			continue
+		}
+
+		pkg := &corev1.AvailablePackage{
+			Name:       entry[0].Name,
+			Version:    entry[0].Version,
+			IconUrl:    entry[0].Icon,
+			Repository: repoRef,
+		}
+		responsePackages = append(responsePackages, pkg)
+	}
+	return responsePackages, nil
+}
+
+func getHelmIndexFileFromURL(indexURL string) (*helmrepo.IndexFile, error) {
+	// Get the response bytes from the url
 	response, err := http.Get(indexURL)
 	if err != nil {
 		return nil, err
 	}
 	defer response.Body.Close()
 
-	if response.StatusCode != 200 {
-		return nil, fmt.Errorf("received non 200 response code: [%d]", response.StatusCode)
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non OK response code: [%d]", response.StatusCode)
```

Comment:
Note, grpc itself will transform any error into a grpc status code (which is then translated into an http status via grpc gateway), so we'll need to be using `status.Errorf` here, rather than `fmt.Errorf`, the former allowing you to specify a status code with the error which can be used for grpc and translated or http.

See https://github.com/kubeapps/kubeapps/blob/master/cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server.go#L95 for an example.

Without doing this, the grpc status will be `codes.Unknown` which is translated to a 500.

More about the grpc status codes at https://grpc.github.io/grpc/core/md_doc_statuscodes.html
And their translation to http in grpc-gateway: https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/errors.go#L34

Reply:
ok will fix it